### ISSUE TITLE
[QUESTION] Regex with XML in request.file and response.file

### ISSUE THREAD

**Issue by itsjxck** [azagniotov/stubby4j#399]
We're using Stubby4J to mock a SOAP API but we're having issues extracting values from the request and injecting them into the response.

The stub config is similar to the following (removed project specific naming):
```
- request:
    method: POST
    url: /path/SomeService
    file: ./requests/ExpectedRequestFormat.xml
 
  response:
    status: 200
    headers:
      content-type: application/xml
    file: ./data/SomeServiceResponse.xml
```

In `ExpectedRequestFormat.xml` we have multiple instances of the following:
```
<ns:SomeElement>${xmlunit.matchesRegex(.*)}</ns:SomeElement>
```

And in `SomeServiceResponse.xml` we have:
```
<ns:SomeOtherElement><% file.1 %></ns:SomeOtherElement>
```

We have also tried the following:
```
<ns:SomeOtherElement><% post.1 %></ns:SomeOtherElement>
<ns:SomeOtherElement><% request.file.1 %></ns:SomeOtherElement>
<ns:SomeOtherElement><% request.post.1 %></ns:SomeOtherElement>
```

However the response is always the raw XML file with no injection of values.

Any ideas what we are doing wrong?

**Comment by azagniotov** [comment]
Hi @itsjxck , 
From a brief glance, based on your examples, I see nothing that jumps out at me.  I will have a deeper look later on today, in ~6-7 hours. Thank you for your patience. 

Also, could you let me know what stubby4j version are you using?

**Comment by azagniotov** [comment]
@itsjxck, I had a look and there two issues:

#### TL;DR
1. Deficient documentation about the behavior of XMLUnit matchers in dynamic token replacement
2. A bug related to token names in dynamic token replacement

#### Details
1. When XMLUnit Matcher is used in the stubbed XML payload, i.e.: `${xmlunit.matchesRegex(.*)}`, this only helps with the XML regex matching. But, using XMLUnit Matcher does not capture the value of the ${xmlunit.matchesRegex(.*)} for token replacement. In other words, currently you can only use dynamic regex token replacement when using vanilla regex syntax (i.e.: `(.*)`), not the XMLUnit matchers. Unfortunately, this is not clear in my docs and this is something I need to clarify better. 
2. In order to use dynamic token replacements for `POST` requests that were stubbed through YAML `file` property, instead of `post`, you have to still use token `<% post.1 %>` instead of `<% file.1 %>`

I made a quick PR #400 to demonstrate that as per above:
1. XMLUnit matching does not work for token replacement
2. When using vanilla regex syntax, `<% post.1 %>` should be used. 

As a recommendation and I realize that it may not be much - if you want to use dynamic token replacement, could you consider not to use `${xmlunit.matchesRegex(.*)}` for now, but just to use vanilla regex synatx?

I will start working on a fix for the above in the meanwhile, thank you for reporting

**Comment by itsjxck** [comment]
@azagniotov thanks for the info!

I'll give it a go with just vanilla matching and your suggestions and will report back any findings.

**Comment by itsjxck** [comment]
Okay so unfortunately it looks like it still doesn't work. I have updated the request file to use vanilla regex like so:
```
<\?xml version="1.0" encoding="utf-8"\?>
...
<ns:SomeElement>(.*)</ns:SomeElement>
```
With the response file updated like so:
```
<ns:OtherElement><% post.1 %></ns:OtherElement>
```
And I still receive the raw xml response file

**Comment by azagniotov** [comment]
That's odd. The test https://github.com/azagniotov/stubby4j/pull/400/files#diff-5411e730423912026ff3af1d740fdce6192bc1cd34b5c065950165ace6a0e3ebR1398-R1425 passes as I am doing what you just described:

YAML config: https://github.com/azagniotov/stubby4j/blob/master/src/functional-test/resources/yaml/include-reported-issues-test-stubs.yaml#L195-L204

1. Stubbed request payload: https://github.com/azagniotov/stubby4j/blob/master/src/functional-test/resources/xml/request/xml_request_issue_399_vanilla_regex_stub_template.xml
2. Stubbed response body: https://github.com/azagniotov/stubby4j/blob/master/src/functional-test/resources/xml/response/xml_response_issue_399_stub_template.xml

I am POSTing to the endpoint: https://github.com/azagniotov/stubby4j/blob/master/src/functional-test/resources/xml/request/xml_request_issue_399_payload.xml and I get back as expected: https://github.com/azagniotov/stubby4j/blob/master/src/functional-test/resources/xml/response/xml_response_issue_399_body.xml

Based on my setup, do you (or am I) do something different?

Also, what stubby4j version are you running?

**Comment by itsjxck** [comment]
I did see those tests and the only difference is the fact the fields we are trying to extract from and inject into are nested. Using version 7.4.0. I unfortunately can't share the complete files but here is the request xml with elements renamed:
```
<\?xml version="1.0" encoding="utf-8"\?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns1="http://example.com/schema1">
  <soap:Body>
    <ns1:RequestElement>
      <ns1:FieldOne>(.*)</ns1:FieldOne>
      <ns1:FieldTwo>(.*)</ns1:FieldTwo>
      <ns1:FieldThree>(.*)</ns1:FieldThree>
      <ns1:FieldFour>(.*)</ns1:FieldFour>
    </ns1:RequestElement>
  </soap:Body>
</soap:Envelope>
```
And the response xml with fields renamed:
```
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns1="http://example.com/schema1">
    <soapenv:Header/>
    <soapenv:Body>
        <ns1:ResponseWrapper>
            <ns1:Response>
                <ns1:Status>SUCCESS</ns1:Status>
                <ns1:Messages>
                  <ns1:Message>
                      <ns1:MessageKey>A Key</ns1:MessageKey>
                      <ns1:MessageText>Some text</ns1:MessageText>
                  </ns1:Message>
                </ns1:Messages>
                <ns1:Collection>
                  <ns1:Data>
                      <ns1:FieldOne><% post.1 %></ns1:FieldOne>
                      <ns1:FieldTwo><% post.2 %></ns1:FieldTwo>
                      <ns1:FieldThree><% post.3 %></ns1:FieldThree>
                      <ns1:FieldFour><% post.4 %></ns1:FieldFour>
                  </ns1:Data>
                </ns1:Collection>
            </ns1:Response>
            <ns1:Snipped/>
        </ns1:ResponseWrapper>
    </soapenv:Body>
</soapenv:Envelope>
```

**Comment by azagniotov** [comment]
Let me take a look, thanks for providing a more complete example

**Comment by itsjxck** [comment]
That really is strange!

1. We are using [this node soap package](https://www.npmjs.com/package/soap) which makes a POST request as confirmed by this line from the Stubby4j logs: `Matched on stubbed METHOD [[POST]] WITH incoming [[POST]]`
2. We don't explicitly set any HTTP headers while using the SOAP client, however I will try this.
3. Stubby4j is running in a docker container using the official image; my env is Ubuntu in WSL2

**Comment by azagniotov** [comment]
Let me do a Docker-based test instead of programmatical test and by making a cURL request to it... What is the stubby Docker image are you running?

**Comment by itsjxck** [comment]
Sorry for the delay, we're using `azagniotov/stubby4j:7.4.0-jre11` specifically

**Comment by azagniotov** [comment]
While I was debugging and trying my own cURL requests against a running standalone JAR (first, before making requests to the JAR running in Docker), I think I actually fixed the original two issues in the last few PRs (in `master` branch now), as described in: https://github.com/azagniotov/stubby4j/issues/399#issuecomment-989611874  , more specifically: using `${xmlunit.matchesRegex(.*)}` to token replacement works both in programmatically tests and when using cURL

#### Stubbed request

```
<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns1="http://example.com/schema1">
<soap:Body>
    <ns1:RequestElement>
        <ns1:FieldOne>${xmlunit.matchesRegex(.*)}</ns1:FieldOne>
        <ns1:FieldTwo>${xmlunit.matchesRegex(.*)}</ns1:FieldTwo>
        <ns1:FieldThree>${xmlunit.matchesRegex(.*)}</ns1:FieldThree>
        <ns1:FieldFour>${xmlunit.matchesRegex(.*)}</ns1:FieldFour>
    </ns1:RequestElement>
</soap:Body>
</soap:Envelope>
```


#### Stubbed response

(Do note that now `<% file.N %>` are used, not `<% post.N %>` like before
```
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns1="http://example.com/schema1">
    <soapenv:Header/>
    <soapenv:Body>
        <ns1:ResponseWrapper>
            <ns1:Response>
                <ns1:Status>SUCCESS</ns1:Status>
                <ns1:Messages>
                    <ns1:Message>
                        <ns1:MessageKey>A Key</ns1:MessageKey>
                        <ns1:MessageText>Some text</ns1:MessageText>
                    </ns1:Message>
                </ns1:Messages>
                <ns1:Collection>
                    <ns1:Data>
                        <ns1:FieldOne><% file.1 %></ns1:FieldOne>
                        <ns1:FieldTwo><% file.2 %></ns1:FieldTwo>
                        <ns1:FieldThree><% file.3 %></ns1:FieldThree>
                        <ns1:FieldFour><% file.4 %></ns1:FieldFour>
                    </ns1:Data>
                </ns1:Collection>
            </ns1:Response>
            <ns1:Snipped/>
        </ns1:ResponseWrapper>
    </soapenv:Body>
</soapenv:Envelope>
```

#### My cURL request
```
curl -d "@issue_399_data.xml" \
  -X POST \
  -H "Content-Type: application/xml" \
  http://localhost:8882/azagniotov/stubby4t/matcher399/xmlunit
```



#### The data file: issue_399_data.xml
```
<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns1="http://example.com/schema1">
<soap:Body>
    <ns1:RequestElement>
        <ns1:FieldOne>banana</ns1:FieldOne>
        <ns1:FieldTwo>apple</ns1:FieldTwo>
        <ns1:FieldThree>strawberry</ns1:FieldThree>
        <ns1:FieldFour>melon</ns1:FieldFour>
    </ns1:RequestElement>
</soap:Body>
</soap:Envelope>
```
#### Expected cURL response with replaced tokens

```
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns1="http://example.com/schema1">
    <soapenv:Header/>
    <soapenv:Body>
        <ns1:ResponseWrapper>
            <ns1:Response>
                <ns1:Status>SUCCESS</ns1:Status>
                <ns1:Messages>
                    <ns1:Message>
                        <ns1:MessageKey>A Key</ns1:MessageKey>
                        <ns1:MessageText>Some text</ns1:MessageText>
                    </ns1:Message>
                </ns1:Messages>
                <ns1:Collection>
                    <ns1:Data>
                        <ns1:FieldOne>banana</ns1:FieldOne>
                        <ns1:FieldTwo>apple</ns1:FieldTwo>
                        <ns1:FieldThree>strawberry</ns1:FieldThree>
                        <ns1:FieldFour>melon</ns1:FieldFour>
                    </ns1:Data>
                </ns1:Collection>
            </ns1:Response>
            <ns1:Snipped/>
        </ns1:ResponseWrapper>
    </soapenv:Body>
</soapenv:Envelope>
```

What I can do now, is push a tag to `azagniotov/stubby4j:latest-jre11` and maybe you could try that?

**Comment by itsjxck** [comment]
Oh interesting! So I can switch back to the xmlunit matching as well? Double win!

I'll give a new version a go once I see it pop up on dockerhub, thanks for investigating this!

**Comment by azagniotov** [comment]
Give me a few mins, I will post a message here when I pushed the tag

**Comment by azagniotov** [comment]
I just pushed to `azagniotov/stubby4j:latest-jre11`. I also checked that the above cURL works against stubby4j running in the Docker container, the response was as expected. 

So to try to sum everything up:
1.  Have a proper formatted XML, so the escapes like this are no longer needed `<\?xml version="1.0"... "\?>` 
2. Try using `${xmlunit.matchesRegex(.*)}`, the dynamic tokens should get replace correctly.
3. Use `<% file.N %>`, not `<% post.N %>` as template tokens

**Comment by itsjxck** [comment]
Perfect! Can confirm that it does indeed work for us as well now!

Thanks for your help with this, really appreciate it!

**Comment by azagniotov** [comment]
Oh, awesome!  I am glad to hear this and I am glad I could help. I will close this issue, please feel free to re-open (or make a new one)

**Comment by azagniotov** [comment]
@itsjxck , I pushed `azagniotov/stubby4j:7.5.0-jre11`, in case if you want to be fixed to a specific tag. It has all the behavior we tested, and some more (all of that was in the `master` already anyways): https://github.com/azagniotov/stubby4j/blob/master/docs/CHANGELOG.md#750

Cheers

**Comment by azagniotov** [comment]
@itsjxck, just FYI:

`stubby4j` Docker images that use `log4j2` and that were affected as a result of [recent log4j2 vulnerability](https://logging.apache.org/log4j/2.x/security.html) (i.e.: `latest` and  <= `v7.5.0`) have been re-built as part of the mitigation effort captured by the following ticket https://github.com/azagniotov/stubby4j/issues/416
